File: server/projects/profile/src/main/java/com/raxim/myscoutee/profile/data/document/mongo/EventItem.java
package com.raxim.myscoutee.profile.data.document.mongo;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.geo.GeoJsonPoint;
import org.springframework.data.mongodb.core.index.GeoSpatialIndexType;
import org.springframework.data.mongodb.core.index.GeoSpatialIndexed;
import org.springframework.data.mongodb.core.mapping.Document;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.raxim.myscoutee.common.repository.GeoJsonPointDeserializer;

@Document(collection = "items")
public class EventItem extends EventBase {
    @Id
    @JsonProperty(value = "key")
    private UUID id;

    // type is general, car, accommodation, private (like phone number)

    // google, booking.com etc., az alkalmazas belso url-je is
    // car is a urlRef - popupwindow with car details
    @JsonProperty(value = "urlRef")
    private String urlRef;

    // opcionalis esemenyek - members

    // Add getters and setters for all fields

    @GeoSpatialIndexed(name = "position", type = GeoSpatialIndexType.GEO_2DSPHERE)
    @JsonDeserialize(using = GeoJsonPointDeserializer.class)
    @JsonProperty(value = "position")
    private GeoJsonPoint position;

    public EventItem() {
    }

    public EventItem(UUID id) {
        this.id = id;
    }

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public String getUrlRef() {
        return urlRef;
    }

    public void setUrlRef(String urlRef) {
        this.urlRef = urlRef;
    }

    public GeoJsonPoint getPosition() {
        return position;
    }

    public void setPosition(GeoJsonPoint position) {
        this.position = position;
    }

    @Override
    public Object clone()
            throws CloneNotSupportedException {
        return super.clone();
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + ((id == null) ? 0 : id.hashCode());
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj == null)
            return false;
        if (getClass() != obj.getClass())
            return false;
        EventItem other = (EventItem) obj;
        if (id == null) {
            if (other.id != null)
                return false;
        } else if (!id.equals(other.id))
            return false;
        return true;
    }

    public void sync() {
        int cnt = (int) getMembers().stream().filter(member -> "A".equals(member.getStatus())).count();
        setNum(cnt);

        if (getMembers() != null && getCapacity() != null) {
            int diff = getCapacity().getMax() - cnt;

            if (diff > 0) {
                getMembers().stream()
                        .filter(member -> "W".equals(member.getStatus()))
                        .sorted((m1, m2) -> m1.getUpdatedDate().compareTo(m2.getUpdatedDate()))
                        .limit(diff).map(member -> {
                            member.setStatus("J");
                            return member;
                        })
                        .collect(Collectors.toList());
            }

            int cntWithJ = (int) getMembers().stream()
                    .filter(member -> "J".equals(member.getStatus())).count();
            int diffWithJ = cntWithJ - getCapacity().getMax();

            if (diffWithJ > 0) {
                getMembers().stream()
                        .filter(member -> "J".equals(member.getStatus()))
                        .sorted((m1, m2) -> m1.getUpdatedDate().compareTo(m2.getUpdatedDate()))
                        .limit(diffWithJ).map(member -> {
                            member.setStatus("W");
                            return member;
                        })
                        .collect(Collectors.toList());
            }
        }
    }
}

File: server/projects/profile/src/main/java/com/raxim/myscoutee/profile/data/document/mongo/EventItem.kt
package com.raxim.myscoutee.profile.data.document.mongo

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.querydsl.core.annotations.QueryEntity
import com.raxim.myscoutee.common.data.converter.GeoJsonPointDeserializer
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.geo.GeoJsonPoint
import org.springframework.data.mongodb.core.index.GeoSpatialIndexType
import org.springframework.data.mongodb.core.index.GeoSpatialIndexed
import org.springframework.data.mongodb.core.mapping.DBRef
import org.springframework.data.mongodb.core.mapping.Document
import org.springframework.format.annotation.DateTimeFormat
import java.util.*

@QueryEntity
@Document(collection = "items")
data class EventItem(
    @Id
    @field:JsonProperty(value = "key")
    val id: UUID? = UUID.randomUUID(),
    //random, main -> only one (there is no random and main -> if all the places has been invited no calculation), general, car, accommodation, private (like phone number)
    //event is published if capacity has been set - auto capacity
    @field:JsonProperty(value = "type")
    val type: String? = null,
    @field:JsonProperty(value = "category")
    val category: String? = null,
    @field:JsonProperty(value = "name")
    val name: String? = null,
    @field:JsonProperty(value = "desc")
    val desc: String? = null,
    // google
    @field:JsonProperty(value = "urlRef")
    val urlRef: String? = null,
    //signal url
    @field:JsonProperty(value = "chatKey")
    val chatKey: String? = null,
    @field:JsonProperty(value = "range")
    var range: RangeLocal? = RangeLocal(),

    @field:JsonProperty(value = "capacity")
    var capacity: RangeInt? = RangeInt(),

    @field:JsonProperty(value = "telNum")
    val telNum: String? = null,
    @field:JsonProperty(value = "optional")
    val optional: Boolean? = null, // not mandatory event - not all participant needs to attend
    @field:JsonProperty(value = "amount")
    val amount: Amount? = Amount(),
    @field:JsonProperty(value = "ticket")
    val ticket: Boolean? = null,

    //@DBRef
    // kivalasztas belemasolas - modositas nem vezetodik at, ha valtozott a kocsi rendszama - update
    @field:JsonProperty(value = "car")
    val car: Car? = null,

    //the two item below is needed for the random calculation, which replaces members, if they are not available

    //when you met with a member
    @field:JsonProperty(value = "memberRange")
    var memberRange: RangeLocal? = RangeLocal(),

    @GeoSpatialIndexed(name = "position", type = GeoSpatialIndexType.GEO_2DSPHERE)
    @JsonDeserialize(using = GeoJsonPointDeserializer::class)
    @field:JsonProperty(value = "position")
    var position: GeoJsonPoint? = null,

    // opcionalis esemenyek - car is egy esemeny, amihez lehet csatlakozni
    // memberek telefonszama nincs itt mentve, egyelore
    @get:JsonIgnore
    @set:JsonProperty(value = "members")
    var members: MutableSet<Member>? = mutableSetOf(),

    @field:JsonProperty(value = "num")
    var num: Int = 0,

    @field:JsonProperty(value = "rule")
    var rule: Rule? = null,

    //Active (A), Deleted (D)
    @field:JsonProperty(value = "status")
    var status: String? = "A",

    @field:JsonProperty(value = "priority")
    val priority: Boolean? = null,

    @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
    @field:JsonProperty(value = "createdDate")
    var createdDate: Date? = Date()
)